File: 백준/Gold/14719. 빗물/빗물.py
import sys
input = sys.stdin.readline

h,w = map(int,input().split())
block = list(map(int, input().split()))

res = 0
left = 0
for i in range(len(block)):
    if block[i]>left:
        left = i
        break
  
while left != (w-1):
    max_block = 0
    right = 0
    
    for i in range(left+1,w):
        if block[i] >= max_block:
            right = i
            max_block = block[i]
            
    if max_block == 0:
        break
         
    if right-left > 1:
        a = min(block[left],block[right])
        left_idx = left

        for j in range(left+1, right):
            
            
            if block[left_idx] < block[j]:
                left_idx = j
            a = min(block[left_idx],block[right])
            
            res += abs(a-block[j])
    
    left = right

print(res)

File: 백준/Gold/14719. 빗물/빗물.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int H = Integer.parseInt(st.nextToken());
        int W = Integer.parseInt(st.nextToken());
        int[] arr = new int[W];
        st = new StringTokenizer(br.readLine());
        for(int i=0; i<W; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int res = 0;
        for(int i=1; i<W-1; i++) {
            int left=0;
            int right=0;

            for(int l=0; l<i; l++) {
                left = Math.max(left, arr[l]);
            }

            for(int r=i+1; r<W; r++) {
                right = Math.max(right, arr[r]);
            }

            if(arr[i]<=left && arr[i]<=right) {
                res += (Math.min(left, right)-arr[i]);
            }
        }
        System.out.println(res);
    }
}
